src/Client/SignatureVerifier.php

```diff
@@ -0,0 +1,140 @@
+        $needVerified = $role->getThreshold();
+        $verifiedKeySignatures = [];
+
+        $canonicalBytes = JsonNormalizer::asNormalizedJson($metadata->getSigned());
+        foreach ($signatures as $signature) {
+            // Don't allow the same key to be counted twice.
+            if ($role->isKeyIdAcceptable($signature['keyid']) && $this->verifySingleSignature($canonicalBytes, $signature)) {
+                $verifiedKeySignatures[$signature['keyid']] = true;
+            }
+            // @todo Determine if we should check all signatures and warn for
+            //     bad signatures even if this method returns TRUE because the
+            //     threshold has been met.
+            //     https://github.com/php-tuf/php-tuf/issues/172
+            if (count($verifiedKeySignatures) >= $needVerified) {
+                break;
+            }
+        }
+
+        if (count($verifiedKeySignatures) < $needVerified) {
+            throw new SignatureThresholdExpception("Signature threshold not met on " . $metadata->getRole());
+        }
+    }
+
+    /**
+     * @param string $bytes
+     *     The canonical JSON string of the 'signed' section of the given file.
+     * @param \ArrayAccess $signatureMeta
+     *     The ArrayAccess object of metadata for the signature. Each signature
+     *     metadata contains two elements:
+     *     - keyid: The identifier of the key signing the role data.
+     *     - sig: The hex-encoded signature of the canonical form of the
+     *       metadata for the role.
+     *
+     * @return boolean
+     *     TRUE if the signature is valid for the.
+     */
+    private function verifySingleSignature(string $bytes, \ArrayAccess $signatureMeta): bool
+    {
+        // Get the pubkey from the key database.
+        $pubkey = $this->keyDb->getKey($signatureMeta['keyid'])->getPublic();
+
+        // Encode the pubkey and signature, and check that the signature is
+        // valid for the given data and pubkey.
+        $pubkeyBytes = hex2bin($pubkey);
+        $sigBytes = hex2bin($signatureMeta['sig']);
+        // @todo Check that the key type in $signatureMeta is ed25519; return
+        //     false if not.
+        //     https://github.com/php-tuf/php-tuf/issues/168
+        return \sodium_crypto_sign_verify_detached($sigBytes, $bytes, $pubkeyBytes);
+    }
+
+    /**
+     * Adds a role to the signature verifier.
+     *
+     * @param \Tuf\Role $role
+     *
+     * @throws \Exception
+     */
+    public function addRole(Role $role):void
+    {
+        if (!$this->roleDb->roleExists($role->getName())) {
+            $this->roleDb->addRole($role);
+        }
+    }
+
+    /**
+     * Adds a key to the signature verifier.
+     *
+     * @param string $keyId
+     * @param \Tuf\Key $key
+     *
+     * @throws \Tuf\Exception\InvalidKeyException
+     */
+    public function addKey(string $keyId, Key $key):void
+    {
+        $this->keyDb->addKey($keyId, $key);
```

Comment:
Does KeyDB check for duplicates?

Reply:
No it doesn't but because the computed key id is checked you could never add key that was different for the same id.

So calling it 2x would not change the DB state